File: r_python_julia/linear_search.jl
# Copyright (c) 2021 Daniel C. Moura
# ** LINK TO ARTICLE GOES HERE **
# linkedin.com/in/dmoura
# twitter.com/daniel_c_moura

# running this file:
# julia --optimize=3 --check-bounds=no linear_search.jl

using CPUTime

### functions under evaluation ###

# 0.37 seconds
in_search(vec, x) = x in vec

# 0.32 seconds
ff_search(vec, x) = findfirst(el -> el == x, vec) != nothing

# 0.38 seconds
mapr_search(vec, x) = mapreduce(el -> el == x, |, vec)

# 1.10 seconds
vec_search(vec, x) = any(vec .== x)

gen_search(vec, x) = !isempty(nothing for v in vec if v == x)

# 0.33 seconds
function foreach_search(vec, x)
    @simd for v in vec
        if v == x
            return true
        end
    end
    false
end

# 0.32 seconds
function for_search(vec, x)
    @inbounds @simd for i in eachindex(vec)
        if vec[i] == x
            return true
        end
    end
    false
end



### aux functions ###

function bench(label, f, v)
    f(v[1:10], 0) #warmup

    println("$label:")
    for r in 1:3 # three runs
        print("\t")
        GC.gc() #cleaning the garbage before each measurement
        @CPUtime begin
            nmatches = 0
            for i = 1:1000 # one thousand searches per run
                if f(v, i)
                    nmatches += 1
                end
            end
        end
        println("\t\tmatches: $nmatches")
    end
    println()
end

# reads a vector of Ints from a text file
function readvec(filename)
    open(filename, "r") do f
        map(line -> parse(Int, line), eachline(f))
    end
end

# reads a vector of Ints from a text file, but the result is
#   a vector of Any... will hurt performance
function readvec_bad(filename)
	a = [] # this is the problem... it should be `a = Int[]`
    open(filename, "r") do f
        for line in eachline(f)
            push!(a, parse(Int, line))
        end
    end
    a
end


### main ###

function main()
    a = readvec("vec.txt")
    #a = readvec_bad("vec.txt")
    println("$(length(a)) records")

    bench("IN",         in_search,      a)
    bench("FINDFIRST",  ff_search,      a)
    bench("MAPR",       mapr_search,    a)
    bench("VEC",        vec_search,     a)
    bench("FOREACH",    foreach_search, a)
    bench("FOR",        for_search,     a)
end

main()

File: r_python_julia/linear_search.c
// Copyright (c) 2021 Daniel C. Moura
// https://towardsdatascience.com/r-vs-python-vs-julia-90456a2bcbab
// linkedin.com/in/dmoura
// twitter.com/daniel_c_moura

// compiling this file:
// gcc linear_search.c -O3

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <time.h>

#define MAX_LINE_LEN 10
#define ARRAY_SIZE 1000000

// 0.26 seconds
int for_search(const int *vec, int size, int x) {
  for (int i=0;i<size;i++)
    if (vec[i] == x)
      return 1;
  return 0;
}

int main() {
  char *filename = "vec.txt";
  FILE *f;
  char line[MAX_LINE_LEN];
  int vec[ARRAY_SIZE];
  clock_t start, end;
  double elapsed_used;

  //simple file reading with predefined number of lines
  f = fopen(filename, "r");
  int i = 0;
  while (fgets(line, MAX_LINE_LEN, f) != NULL)
    vec[i++] = (int) strtoul(line, NULL, 0);
  fclose(f);
  printf("%d records\n", i);

  for (int r=0;r<3;r++) { // three runs        
    start = clock();

    int nmatches = 0;
    for (int i=1;i<=1000;i++)
      if (for_search(vec, ARRAY_SIZE, i))
        nmatches++;

    end = clock();
    elapsed_used = ((double)(end - start)) / CLOCKS_PER_SEC;

    printf("elapsed CPU time: %f seconds\n", elapsed_used);
    printf("\tmatches: %d\n", nmatches);
  }
  return 0;
}
